mlir/lib/Dialect/MemRef/Transforms/MergeAllocTickBased.cpp

```diff
@@ -0,0 +1,383 @@
+  while (!s->complexScopeStack.empty()) {
+    auto &scope = s->complexScopeStack.back();
+    if (!op || !scope.scope->isProperAncestor(op)) {
+      if (failed(onPopComplexScope(s, s->curTick))) {
+        return failure();
+      }
+      s->complexScopeStack.pop_back();
+    } else {
+      break;
+    }
+  }
+  return success();
+}
+
+void TickCollecter::forwardTick(TickCollecterStates *s) const { s->curTick++; }
+
+void TickCollecter::accessValue(TickCollecterStates *s, Value v,
+                                bool complex) const {
+  if (auto refv = dyn_cast<TypedValue<MemRefType>>(v)) {
+    for (auto &&base : s->aliasAnaly.resolveReverse(refv)) {
+      auto defop = base.getDefiningOp();
+      if (isa_and_present<memref::AllocOp>(defop)) {
+        s->allocTicks[defop].update(complex ? UNTRACEABLE_ACCESS : s->curTick);
+        if (!s->complexScopeStack.empty()) {
+          s->complexScopeStack.back().operations.insert(defop);
+        }
+      }
+    }
+  }
+}
+
+void TickCollecter::onMemrefViews(TickCollecterStates *s,
+                                  ViewLikeOpInterface op) const {
+  auto viewSrc = op.getViewSource();
+  // don't need to access the first operand, which is "source".
+  // The "source" operand is not really read or written at this point
+  for (auto val : op.getOperation()->getOperands()) {
+    if (val != viewSrc)
+      accessValue(s, val, false);
+  }
+}
+
+void TickCollecter::onReturnOp(TickCollecterStates *s, Operation *op) const {
+  bool isTopLevel = isa<func::FuncOp>(op->getParentOp());
+  for (auto val : op->getOperands()) {
+    accessValue(s, val, isTopLevel);
+  }
+}
+
+void TickCollecter::onAllocOp(TickCollecterStates *s, Operation *op) const {
+  s->allocTicks[op].allocTick = s->curTick;
+}
+
+void TickCollecter::onGeneralOp(TickCollecterStates *s, Operation *op) const {
+  for (auto val : op->getOperands()) {
+    accessValue(s, val, false);
+  }
+}
+
+void TickCollecter::pushComplexScope(TickCollecterStates *s,
+                                     Operation *op) const {
+  s->complexScopeStack.emplace_back(op, s->curTick);
+}
+
+bool TickCollecter::isMergeableAlloc(TickCollecterStates *s, Operation *op,
+                                     int64_t tick) const {
+  if (tick == UNTRACEABLE_ACCESS) {
+    return false;
+  }
+  if (!isMemRefTypeOk(cast<MemRefType>(op->getResultTypes().front()))) {
+    return false;
+  }
+  auto alignment = cast<memref::AllocOp>(op).getAlignment();
+  if (!alignment) {
+    return true; // ok if no alignment
+  }
+  return alignment > 0 && (s->opt.alignment % alignment.value() == 0);
+}
+
+// find the closest surrounding parent operation with AutomaticAllocationScope
+// trait, and is not scf.for
+Operation *TickCollecter::getAllocScope(TickCollecterStates *s,
+                                        Operation *op) const {
+  auto parent = op;
+  for (;;) {
+    parent = parent->getParentWithTrait<OpTrait::AutomaticAllocationScope>();
```

Comment:
Yes, `scf.if` is not a `AutomaticAllocationScope`, the pass will find the `scf.forall` instead

Reply:
We might need special handle as well for the allocation happened inside `scf.if`? as it might not always true that we need to allocate this buffer, right?

Reply:
Yes, I agree. `scf.if` is `RegionBranchOpInterface` which we also take care of. :)